fix: adds check for null model to componentWillReceiveProps

logandavis opened the pull request:
There are some cases where a Backbone model passed through CBR
might become undefined. (Consider the case where a view relies on
a secondary model as a "decorator", and the use of that "decorator"
model is toggled on or off.) Today, when a model becomes undefined,
CBR throws a TypeError in componentWillReceiveProps:

Uncaught TypeError: Cannot read property 'on' of undefined

This commit adds a check for null or undefined models before trying
to set event listeners in componentWillReceiveProps.

logandavis committed:
fix: adds check for null model to componentWillReceiveProps

There are some cases where a Backbone model passed through CBR
might become undefined. (Consider the case where a view relies on
a secondary model as a "decorator", and the use of that "decorator"
model is toggled on or off.) Today, when a model becomes undefined,
CBR throws a TypeError in ComponentWillReceiveProps:

Uncaught TypeError: Cannot read property 'on' of undefined

This commit adds a check for null or undefined models before trying
to set event listeners in componentWillReceiveProps.

logandavis committed:
Removes listeners from models which become undefined

logandavis commented:
cc @hswolff -- curious what you think!

hswolff commented on lib/connect-backbone-to-react.js:
```diff
@@ -128,6 +128,18 @@ module.exports = function connectBackboneToReact(
+            // Instead, if it was previously defined, remove the old listeners.
+            const prevModel = this.props.models[mapKey];
+            if (prevModel) {
+              getEventNames(mapKey).forEach(name => {
```
This is duped now, let's create a util method for this functionality. `removeListeners(mapKey, model) {}`.

hswolff commented on lib/connect-backbone-to-react.js:
```diff
@@ -128,6 +128,18 @@ module.exports = function connectBackboneToReact(
+          // Do not attempt to create event listeners on an undefined model.
+          if (!model) {
+            // Instead, if it was previously defined, remove the old listeners.
+            const prevModel = this.props.models[mapKey];
```
Should we check `this.context` as well?

hswolff commented on lib/connect-backbone-to-react.js:
```diff
@@ -128,6 +128,18 @@ module.exports = function connectBackboneToReact(
             (this.context.models && this.context.models[mapKey] === this.models[mapKey])) {
```
Also just realized that we're duplicating `this.models[mapKey]` when we have a reference to `model` above. Can we squeeze in a change to reference `model` directly for some code hygiene?

hswolff reviewed (changes requested):
Can we add a test case for this as well please?

logandavis committed:
DRYs removeEventListener code

logandavis committed:
Reference model directly in componentWillReceiveProps

logandavis committed:
Checks this.context.models for models becoming undefined

Previously, we only checked this.props.models for models becoming
undefined in componentWillReceiveProps. This commit also checks
this.context.models, but only if a matching model was not found in
this.props.models -- because of the Object.assign in setModels,
if there are two models with the same mapKey, we prefer the one
in props.

logandavis committed:
Adds test for model becoming undefined in componentWillReceiveProps

logandavis commented:
Updated with a test case -- thanks for the review! cc @hswolff

hswolff commented on lib/connect-backbone-to-react.js:
```diff
@@ -124,11 +130,25 @@ module.exports = function connectBackboneToReact(
           }
 
+          // Do not attempt to create event listeners on an undefined model.
+          if (!model) {
```
Should this block come before the first if block in this function?

Also we're looking up the propModel and the contextModel twice. Every time you try and access a property on an object you have a small hit to performance. Putting aside the micro-optimization, it may make this code easier to read if we cache that lookup at the top of the block so we don't have to repeat the boilerplate code in two locations.

logandavis replied:
Yup, cleaned this up a bit 🙂

hswolff commented on test/connect-backbone-to-react.test.js:
```diff
@@ -594,6 +594,60 @@ describe('connectBackboneToReact', function() {
+      createListenerSpy = sandbox.spy(ConnectedTest.prototype, 'createEventListener');
+      removeListenerSpy = sandbox.spy(ConnectedTest.prototype, 'removeEventListener');
+
+      const newModelsMap = {
```
This should be done in a nested beforeEach block. Such that we can verify the known state before we then update to the next state that we're testing.

logandavis committed:
Micro-optimize and clean componentWillReceiveProps

logandavis committed:
Refactor undefined model test per review

logandavis commented on test/connect-backbone-to-react.test.js:
```diff
@@ -594,6 +594,65 @@ describe('connectBackboneToReact', function() {
+      let createListenerSpy;
+      let removeListenerSpy;
+
+      beforeEach(function() {
```
@hswolff Is this what you meant by moving the setup to a nested beforeEach?

hswolff replied:
Yep! But let's add come assertions before the describe block to assert that the event listeners are initially bound, bound once, etc. Such that we can then assert they are removed as well. It's a little redundant, but I hope more readable.

logandavis commented:
Updated cc @hswolff

hswolff commented on lib/connect-backbone-to-react.js:
```diff
@@ -124,11 +130,25 @@ module.exports = function connectBackboneToReact(
+            if (propsModel) {
+              this.removeEventListener(mapKey, propsModel);
+            } else if (contextModel) {
+              // this.setModel prefers props to context, so we do that here too.
```
Outdated comment?

logandavis replied:
Ah no -- the comment refers to the `else if` right above it. It's an attempt to explain the potentially surprising behavior that if a model with the matching mapKey exists in both `props` and `context`, we only remove listeners from the one in `props`. We do this because only the one in `props` is actually used in `this.models`, per the `Object.assign` in `setModel`.

Let me know if you think there's a clearer way to express this?

hswolff replied:
Ah.....how about lets use your wording? Put this above the if statement.

>  if a model with the matching mapKey exists in both props and context, we only remove listeners from the one in props. 
> We do this because only the one in props is actually used in this.models, per the Object.assign in setModel.

hswolff reviewed (changes requested):
Looking really good! Love that refactor! 2 last small comments.

logandavis committed:
Adds testing for initial model listener bindings

logandavis commented:
Updated cc @hswolff

hswolff commented on test/connect-backbone-to-react.test.js:
```diff
@@ -594,6 +594,76 @@ describe('connectBackboneToReact', function() {
+    });
+
+    it('calls createEventListener once for the model', function() {
+      assert.equal(createListenerSpy.callCount, 1);
```
How is this only called once? Don't we have 3 Models/Collections? Shouldn't this be called 3 times? I'm a little confused.

logandavis replied:
It's only called once because the userModel and userCollection haven't changed between the mounting of the view [here](https://github.com/mongodb-js/connect-backbone-to-react/pull/21/files#diff-f93d5182925707305633f3d1d9f6a73bR605) and the props update [here](https://github.com/mongodb-js/connect-backbone-to-react/pull/21/files#diff-f93d5182925707305633f3d1d9f6a73bR623). Since only the decorator has changed since the listeners were set up, createEventListener has only been called for the decorator.

hswolff replied:
`calls createEventListener once due to decoratorUserModel being added as a model`?

hswolff commented on test/connect-backbone-to-react.test.js:
```diff
@@ -594,6 +594,76 @@ describe('connectBackboneToReact', function() {
+        assert.equal(createListenerSpy.callCount, 1);
+      });
+
+      it('calls removeEventListener once for the model', function() {
```
We should assert that it's being called for the model that we are removing, as that's the behavior we're concerned about.

logandavis replied:
Good idea!

logandavis committed:
Adds tests: createListener and removeListener are called for decorator model

logandavis commented:
cc @hswolff I think that's the last of the functional changes we discussed, but please let me know if there's anything you'd find more readable! In particular I think there are some outstanding readability concerns about [this comment](https://github.com/mongodb-js/connect-backbone-to-react/pull/21#discussion_r227563818) and the wording of [this test](https://github.com/mongodb-js/connect-backbone-to-react/pull/21#discussion_r227805876).

hswolff commented:
Last two updates and we should be good to go.

logandavis committed:
Updates test titles and comments per review

logandavis commented:
cc @hswolff Comments and test titles updated!

hswolff commented:
Thank you so much!

Published as `v1.6.1` 🎆
